ddtrace/tracer/logger.go

```diff
@@ -0,0 +1,42 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016 Datadog, Inc.
+
+package tracer
+
+import "github.com/DataDog/dd-trace-go/v2/internal/log"
+
+// Logger implementations are able to log given messages that the tracer or profiler might output.
+type Logger interface {
+	// Log prints the given message.
+	Log(msg string)
+}
+
+// UseLogger sets l as the logger for all tracer and profiler logs.
+func UseLogger(l Logger) {
+	log.UseLogger(l)
+}
+
+// LogLevel represents the logging level that the log package prints at.
+type LogLevel log.Level
+
+func (l LogLevel) String() string {
+	return log.Level(l).String()
+}
+
+type loggerAdapter struct {
+	fn func(lvl LogLevel, msg string, a ...any)
+}
+
+func (l loggerAdapter) Log(msg string) {
+	l.fn(LogLevel(log.CurrentLevel()), msg)
```

Comment:
🐞 I don't think this works. `CurrentLevel` isn't the same as the `lvl` argument passed to `printMsg` in `internal/log/log.go`.

<details>
<summary>Proof</summary>

**Test Case**

```go
func TestAdaptLogger(t *testing.T) {
	l := AdaptLogger(func(lvl LogLevel, msg string, a ...any) {
		fmt.Printf("level=%s: %s", lvl, msg)
	})
	UseLogger(l)

	log.Warn("warning\n")
	log.Info("info\n")
}
```

**Output**

```bash
$ go test -run 'Adapt' -v ./ddtrace/tracer/
=== RUN   TestAdaptLogger
level=WARN: Datadog Tracer v2.0.0-dev WARN: warning
level=WARN: Datadog Tracer v2.0.0-dev INFO: info
```

Note `level=WARN` for both messages. The second one should be `level=INFO`.

</details>

Reply:
@felixge You were right. I pushed a new implementation that relies in an anonymous interface.